provisioning/transport/amqp/src/SaslTpmHandler.cs

```diff
@@ -0,0 +1,183 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Text;
+using Microsoft.Azure.Amqp;
+using Microsoft.Azure.Amqp.Sasl;
+using Microsoft.Azure.Devices.Shared;
+
+namespace Microsoft.Azure.Devices.Provisioning.Client.Transport.Amqp
+{
+    internal class SaslTpmHandler : SaslHandler
+    {
+        private const string MechanismName = "TPM";
+        private const int ChallengeKeySegmentSize = 400;
+
+        private readonly byte[] _endorsementKey;
+        private readonly byte[] _storageRootKey;
+
+        private readonly string _hostName;
+        private readonly StringBuilder _encodedNonceStringBuilder = new StringBuilder(ChallengeKeySegmentSize * 3);
+        private readonly ProvisioningSecurityClientSasToken _security;
+        private byte _nextSequenceNumber;
+
+        public SaslTpmHandler(byte[] endorsementKey, byte[] storageRootKey, string hostName, ProvisioningSecurityClientSasToken security)
+
+        {
+            //TODO: check inputs
+
+            Mechanism = MechanismName;
+            _endorsementKey = endorsementKey;
+            _storageRootKey = storageRootKey;
+            _hostName = hostName;
+            _security = security;
+        }
+
+        public override SaslHandler Clone()
+        {
+            return new SaslTpmHandler(_endorsementKey, _storageRootKey, _hostName, _security);
+        }
+
+        public override void OnChallenge(SaslChallenge challenge)
+        {
+            var challengeAction = GetChallengeAction(challenge);
+
+            switch (challengeAction)
+            {
+                case SaslChallengeAction.First:
+                    SendStorageRootKey();
+                    break;
+                case SaslChallengeAction.Interim:
+                    SaveEncodedNonceSegment(challenge);
+                    SendInterimResponse();
+                    break;
+                case SaslChallengeAction.Final:
+                    SaveEncodedNonceSegment(challenge);
+                    SendLastResponse();
+                    break;
+            }
+        }
+
+        private void SendInterimResponse()
+        {
+            var response = new SaslResponse {Response = new ArraySegment<byte>(new byte[1] {0})};
```

Comment:
Please add a constant if this single byte is never changed or read otherwise this is going to use a lot of memory.